tools/build_langserver/langserver_main.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"context"
+	"gopkg.in/op/go-logging.v1"
+	"os"
+
+	"cli"
+	"github.com/sourcegraph/jsonrpc2"
+	"tools/build_langserver/langserver"
+)
+
+// TODO(bnmetrics): also think about how we can implement this with .build_defs as well
+// TODO(bnmetrics): Make the Usage part better
+var log = logging.MustGetLogger("build_langserver")
+
+var opts = struct {
+	Usage     string
+	Verbosity cli.Verbosity `short:"v" long:"verbosity" default:"warning" description:"Verbosity of output (higher number = more output)"`
+
+	Mode string `short:"m" long:"mode" default:"stdio" description:"Mode of the language server communication (stdio|tcp)"`
+	Host string `short:"h" long:"host" default:"127.0.0.20" description:"TCP host to communicate with"`
+	Port string `short:"p" long:"port" default:"4387" description:"TCP port to communicate with"`
+}{
+	Usage: `
+build_langserver is a binary shipped with Please that you can use as a language server for build files.
+
+It speaks language server protocol from vscode, you can plugin this binary in your IDE to start the language server.
+Currently, it supports autocompletion, goto definition for build_defs, and signature help
+`,
+}
+
+func main() {
+	cli.ParseFlagsOrDie("build_langserver", "1.0.0", &opts)
+	cli.InitLogging(opts.Verbosity)
+
+	handler := langserver.NewHandler()
+
+	serve(handler)
+}
+
+func serve(handler jsonrpc2.Handler) {
+	switch opts.Mode {
+	case "stdio":
+		// TODO(bnmetrics): complete this
+	case "tcp":
```

Comment:
aren't these the wrong way around? isn't this doing stdin/stdout?

Also if you do the flags thing, you can probably skip the switch and default case and just do
```
if opts.Mode == "tcp" {
} else {
    // stdio stuff
}
```
because you know it's always one of the two.

Reply:
Yes! I moved things around and got things out of order..oops!
the flag actually take care of the invalid options? Nice!